roles/bootstrap-os/tasks/main.yml

```diff
@@ -21,9 +21,13 @@
 - name: Gather nodes hostnames
   setup:
     gather_subset: '!all'
-    filter: ansible_hostname
+    filter: ansible_*
 
 - name: Assign inventory name to unconfigured hostnames
-  hostname:
-    name: "{{inventory_hostname}}"
+  block:
```

Comment:
First of all,we  are not in a trend to use block: expression. 
Secondly, using hostnamectl does not update ansible fact for the hostname. That's the whole point behind using hostname function. Does it not work correctly in CoreOS?

Reply:
1. I don't want to be a trendsetter :-) so let's flatten it out
2. Yeah right, I did not think about that sideeffect. However I tried CoreOS 1409 and 1298 and had no luck. I would have expected that the `hostname` module selects the `SystemdStrategy`  [here](https://github.com/ansible/ansible-modules-core/blob/devel/system/hostname.py#L115-L118):

```python
        if self.platform == 'Linux' and Facts(module).is_systemd_managed():
            self.strategy = SystemdStrategy(module)
        else:
            self.strategy = self.strategy_class(module)
```
I'm no Ansible expert - is the behavior I see somehow related to the defered facts gathering?

Reply:
ok so I dug deeper and tried several approaches for boostrapping python on CoreOS. I believe the problem is that python's `platform.linux_distribution()` does not return anything usable ob CoreOS. However, the `hostname` module uses `load_platform_subclass` to determine the strategy to use, and [load_platform_subclass()](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L317-L340) uses `get_distribution` (which in turn uses `platform.linux_distribution()`) to do that in [basic.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L323-L324)

I tried both `pypy` and `ActivePython`, and for both `platform.linux_distribution()` does not work:

```bash
core@localhost ~ $ cat /etc/os-release
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1298.7.0
VERSION_ID=1298.7.0
BUILD_ID=2017-03-31-0215
...
core@localhost ~ $ python
Python 2.7.13 (c925e7381036, Jun 05 2017, 21:20:51)
[PyPy 5.8.0 with GCC 6.2.0 20160901] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>>> import platform
>>>> print("Platform: %s, Distro: %s" % (platform.system(), platform.linux_distribution()[0]))
Platform: Linux, Distro:
>>>>
```

Reply:
@olmoser Thanks for your careful research. Let's do this in a slightly different way:

```
- name: Gather nodes hostnames 
   setup:
     gather_subset: '!all'
     filter: ansible_hostname, ansible_os_*

- name: Assign inventory name to unconfigured hostnames (non-CoreOS)
  hostname:
    name: "{{inventory_hostname}}"
  when: ansible_os_family not in ['CoreOS', 'Container Linux by CoreOS']

- name: Assign inventory name to unconfigured hostnames (CoreOS only)
  command: "hostnamectl set-hostname  {{inventory_hostname}}"
  register: hostname_changed
  when: ansible_hostname == 'localhost' and ansible_os_family in ['CoreOS', 'Container Linux by CoreOS']

- name: Update hostname fact (CoreOS only)
   setup:
     gather_subset: '!all'
     filter: ansible_hostname, ansible_os_*
  when: ansible_os_family in ['CoreOS', 'Container Linux by CoreOS'] and hostname_changed.changed
```

Reply:
@mattymo great, thanks for the writeup, that's anyways pretty much what I ended up with after removing the block expression. I wasn't aware that you can use multiple `filter` rules in the `setup` block mit comma delimiting them, good to know. I'll adapt the PR.